**Import du dataset depuis hunginface**

In [2]:
import os
import urllib.request

os.makedirs("medical_bios_data", exist_ok=True)

url = "https://huggingface.co/datasets/coastalcph/medical-bios/resolve/main/bios.zip"
dest = "medical_bios_data/bios.zip"

if not os.path.exists(dest):
    print("Telechargement de bios.zip...")
    urllib.request.urlretrieve(url, dest)
    print("Telechargement termine.")
else:
    print("bios.zip deja present, telechargement ignore.")


bios.zip deja present, telechargement ignore.


In [64]:
# Étape 4a: Extraire le fichier zip
import zipfile
import os

zip_path = 'medical_bios_data/bios.zip'
extract_path = 'medical_bios_data/'

# Vérifier que le zip existe 
if os.path.exists(zip_path):
    print(f" Fichier zip trouvé: {zip_path}")
    print(f" Taille: {os.path.getsize(zip_path) / (1024*1024):.2f} MB")
    
    # Extraire le contenu
    with zipfile.ZipFile(zip_path, 'r') as zip_ref:
        zip_ref.extractall(extract_path)
        print(" Extraction terminée!")
        
        # Lister les fichiers dans le zip
        print("\n Contenu du zip:")
        for file_info in zip_ref.infolist():
            print(f"  - {file_info.filename} ({file_info.file_size} bytes)")
            
else:
    print(f" Fichier zip non trouvé: {zip_path}")

 Fichier zip trouvé: medical_bios_data/bios.zip
 Taille: 1.82 MB
 Extraction terminée!

 Contenu du zip:
  - test_rationales.jsonl (300522 bytes)
  - __MACOSX/._test_rationales.jsonl (176 bytes)
  - test.jsonl (946345 bytes)
  - __MACOSX/._test.jsonl (275 bytes)
  - train.jsonl (7595575 bytes)
  - __MACOSX/._train.jsonl (219 bytes)
  - validation.jsonl (938592 bytes)
  - __MACOSX/._validation.jsonl (219 bytes)


In [65]:
# Étape 4b: Vérifier le contenu extrait
import os

# Lister tous les fichiers dans le dossier medical_bios_data
print(" Contenu du dossier medical_bios_data/:")
for root, dirs, files in os.walk('medical_bios_data/'):
    for file in files:
        file_path = os.path.join(root, file)
        file_size = os.path.getsize(file_path)
        print(f"   {file_path} ({file_size:,} bytes)")

# Vérifier spécifiquement les fichiers JSONL
jsonl_files = ['train.jsonl', 'validation.jsonl', 'test.jsonl']
print(f"\n Recherche des fichiers JSONL:")

for jsonl_file in jsonl_files:
    # Chercher dans le dossier et sous-dossiers
    found = False
    for root, dirs, files in os.walk('medical_bios_data/'):
        if jsonl_file in files:
            full_path = os.path.join(root, jsonl_file)
            size = os.path.getsize(full_path)
            print(f"   {full_path} ({size:,} bytes)")
            found = True
            break
    if not found:
        print(f"   {jsonl_file} non trouvé")

 Contenu du dossier medical_bios_data/:
   medical_bios_data/bios.zip (1,912,483 bytes)
   medical_bios_data/test.jsonl (946,345 bytes)
   medical_bios_data/test_rationales.jsonl (300,522 bytes)
   medical_bios_data/train.jsonl (7,595,575 bytes)
   medical_bios_data/validation.jsonl (938,592 bytes)
   medical_bios_data/__MACOSX\._test.jsonl (275 bytes)
   medical_bios_data/__MACOSX\._test_rationales.jsonl (176 bytes)
   medical_bios_data/__MACOSX\._train.jsonl (219 bytes)
   medical_bios_data/__MACOSX\._validation.jsonl (219 bytes)

 Recherche des fichiers JSONL:
   medical_bios_data/train.jsonl (7,595,575 bytes)
   medical_bios_data/validation.jsonl (938,592 bytes)
   medical_bios_data/test.jsonl (946,345 bytes)


**chargement des datasets**

In [3]:
# Étape 5: Charger les données JSONL et explorer leur structure
import json
import pandas as pd

def load_jsonl(file_path):
    """Charger un fichier JSONL en liste de dictionnaires"""
    data = []
    with open(file_path, 'r', encoding='utf-8') as f:
        for line in f:
            if line.strip():  # Ignorer les lignes vides
                data.append(json.loads(line))
    return data

def find_file(filename, search_path='medical_bios_data/'):
    """Chercher un fichier récursivement dans un dossier"""
    for root, dirs, files in os.walk(search_path):
        if filename in files:
            return os.path.join(root, filename)
    return None

# Charger les datasets
print(" Chargement des datasets...")

# Trouver le chemin réel de train.jsonl
train_path = find_file('train.jsonl')
print(f" Chemin trouvé: {train_path}")

# Commencer par charger quelques exemples du train pour voir la structure
train_sample = []
with open(train_path, 'r', encoding='utf-8') as f:
    for i, line in enumerate(f):
        if i >= 5:  # Charger seulement les 5 premiers pour l'exploration
            break
        if line.strip():
            train_sample.append(json.loads(line))

print(f" Échantillon chargé: {len(train_sample)} exemples")
print(f"\n Structure d'un exemple:")
print(json.dumps(train_sample[0], indent=2, ensure_ascii=False))

# Regarder les clés disponibles
if train_sample:
    keys = train_sample[0].keys()
    print(f"\n Clés disponibles: {list(keys)}")

 Chargement des datasets...
 Chemin trouvé: medical_bios_data/train.jsonl
 Échantillon chargé: 5 exemples

 Structure d'un exemple:
{
  "full_text": "Dr. Vikram Prasad is an experienced Dentist in Sowkhya Ayurveda Speciality Clinic, Chennai. He has been a practicing Dentist for 20 years. He has done BDS . He is currently associated with Sree Sai Dental Clinic in Sowkhya Ayurveda Speciality Clinic, Chennai. Book an appointment online with Dr. Vikram Prasad and consult privately on Lybrate.com.",
  "text": "He has been a practicing Dentist for 20 years. He has done BDS . He is currently associated with Sree Sai Dental Clinic in Sowkhya Ayurveda Speciality Clinic, Chennai. Book an appointment online with Dr. Vikram Prasad and consult privately on Lybrate.com.",
  "title": "dentist",
  "gender": "Male",
  "date": "2018-09"
}

 Clés disponibles: ['full_text', 'text', 'title', 'gender', 'date']


In [4]:
#chargement du train_set avec la fonction load_jsonl
train_set = load_jsonl('medical_bios_data/train.jsonl')
print(f" Train set chargé: {len(train_set)} exemples") 

 Train set chargé: 8000 exemples


In [5]:
# Étape 6a: Analyse de la complétude des données
import pandas as pd

# Convertir en DataFrame pour l'analyse
df_train = pd.DataFrame(train_set)

print(" ANALYSE DE LA QUALITÉ DES DONNÉES")
print("=" * 50)

# 1. Vérifier les valeurs manquantes
print(" 1. VALEURS MANQUANTES:")
missing_analysis = {}
for column in ['full_text', 'text', 'title', 'gender', 'date']:
    # Compter les valeurs None, vides, ou NaN
    null_count = df_train[column].isnull().sum()
    empty_count = (df_train[column] == '').sum()
    total_missing = null_count + empty_count
    percentage = (total_missing / len(df_train)) * 100
    
    missing_analysis[column] = {
        'null': null_count,
        'empty': empty_count, 
        'total_missing': total_missing,
        'percentage': percentage
    }
    
    status = "" if total_missing == 0 else ""
    print(f"  {status} {column}: {total_missing}/{len(df_train)} manquants ({percentage:.2f}%)")

print(f"\n Taille du dataset: {len(df_train)} exemples")
print(f" Colonnes: {list(df_train.columns)}")

 ANALYSE DE LA QUALITÉ DES DONNÉES
 1. VALEURS MANQUANTES:
   full_text: 0/8000 manquants (0.00%)
   text: 5/8000 manquants (0.06%)
   title: 0/8000 manquants (0.00%)
   gender: 0/8000 manquants (0.00%)
   date: 0/8000 manquants (0.00%)

 Taille du dataset: 8000 exemples
 Colonnes: ['full_text', 'text', 'title', 'gender', 'date']


In [69]:
# Étape 6b: Analyse de la distribution des valeurs
print("\n 2. DISTRIBUTION DES VALEURS:")
print("-" * 30)

# Distribution des genres
print(" GENRE:")
gender_counts = df_train['gender'].value_counts()
print(gender_counts)
print(f"Pourcentages: {df_train['gender'].value_counts(normalize=True) * 100}")

# Distribution des titres/professions
print(f"\n TITRES/PROFESSIONS (Top 10):")
title_counts = df_train['title'].value_counts()
print(title_counts.head(10))

# Analyse des dates
print(f"\n FORMAT DES DATES:")
date_sample = df_train['date'].head(10).tolist()
print(f"Échantillon: {date_sample}")
print(f"Dates uniques: {df_train['date'].nunique()}")

# Longueur des textes
print(f"\n LONGUEUR DES TEXTES:")
df_train['text_length'] = df_train['text'].str.len()
df_train['full_text_length'] = df_train['full_text'].str.len()

print(f"Texte court - Moyenne: {df_train['text_length'].mean():.1f} chars")
print(f"Texte court - Min/Max: {df_train['text_length'].min()} / {df_train['text_length'].max()}")
print(f"Texte complet - Moyenne: {df_train['full_text_length'].mean():.1f} chars") 
print(f"Texte complet - Min/Max: {df_train['full_text_length'].min()} / {df_train['full_text_length'].max()}")


 2. DISTRIBUTION DES VALEURS:
------------------------------
 GENRE:
gender
Female    4290
Male      3710
Name: count, dtype: int64
Pourcentages: gender
Female    53.625
Male      46.375
Name: proportion, dtype: float64

 TITRES/PROFESSIONS (Top 10):
title
psychologist    2200
nurse           1638
dentist         1533
physician       1349
surgeon         1280
Name: count, dtype: int64

 FORMAT DES DATES:
Échantillon: ['2018-09', '2017-39', '2018-09', '2017-43', '2018-09', '2018-09', '2018-09', '2017-43', '2018-09', '2018-09']
Dates uniques: 6

 LONGUEUR DES TEXTES:
Texte court - Moyenne: 375.2 chars
Texte court - Min/Max: 0 / 977
Texte complet - Moyenne: 481.1 chars
Texte complet - Min/Max: 190 / 1000


### Construction du modèle

In [6]:
#chargement des données de validation et test
validation_set = load_jsonl('medical_bios_data/validation.jsonl')
print(f" Validation set chargé: {len(validation_set)} exemples")
test_set = load_jsonl('medical_bios_data/test.jsonl')
print(f" Test set chargé: {len(test_set)} exemples")

 Validation set chargé: 1000 exemples
 Test set chargé: 1000 exemples


In [7]:
# Étape 7: Préparation des données pour l'embedding (sans split)
from sklearn.preprocessing import LabelEncoder
import pandas as pd
import numpy as np

print(" PRÉPARATION DES DONNÉES POUR L'EMBEDDING")
print("=" * 50)

# 1. Préparer tous les datasets en DataFrames de validation et test
df_val = pd.DataFrame(validation_set) 
df_test = pd.DataFrame(test_set)

print(f" TAILLES DES DATASETS:")
print(f"  Train: {len(df_train)} exemples")
print(f"  Validation: {len(df_val)} exemples") 
print(f"  Test: {len(df_test)} exemples")

# 2. Encoder les labels sur TOUT le dataset (train+val+test combined)
le = LabelEncoder()
all_titles = list(df_train['title']) + list(df_val['title']) + list(df_test['title'])
le.fit(all_titles)

# Appliquer l'encodage sur chaque dataset
df_train['title_encoded'] = le.transform(df_train['title'])
df_val['title_encoded'] = le.transform(df_val['title'])
df_test['title_encoded'] = le.transform(df_test['title'])

# 3. Afficher l'encodage des labels
print("\n ENCODAGE DES LABELS:")
label_mapping = dict(zip(le.classes_, le.transform(le.classes_)))
for profession, code in label_mapping.items():
    train_count = (df_train['title'] == profession).sum()
    val_count = (df_val['title'] == profession).sum()
    test_count = (df_test['title'] == profession).sum()
    print(f"  {code}: {profession}")
    print(f"     Train: {train_count}, Val: {val_count}, Test: {test_count}")

# 4. Préparer les données pour l'embedding
print(f"\n PRÉPARATION POUR L'EMBEDDING:")
print(f"  Texte d'entrée: 'full_text'")
print(f"  Labels: {len(le.classes_)} classes")
print(f"  Exemple de texte: '{df_train['full_text'].iloc[0][:100]}...'")

# Sauvegarder le LabelEncoder pour plus tard
import pickle
with open('label_encoder.pkl', 'wb') as f:
    pickle.dump(le, f)
print(f" LabelEncoder sauvegardé dans 'label_encoder.pkl'")

 PRÉPARATION DES DONNÉES POUR L'EMBEDDING
 TAILLES DES DATASETS:
  Train: 8000 exemples
  Validation: 1000 exemples
  Test: 1000 exemples

 ENCODAGE DES LABELS:
  0: dentist
     Train: 1533, Val: 172, Test: 197
  1: nurse
     Train: 1638, Val: 216, Test: 195
  2: physician
     Train: 1349, Val: 172, Test: 170
  3: psychologist
     Train: 2200, Val: 294, Test: 279
  4: surgeon
     Train: 1280, Val: 146, Test: 159

 PRÉPARATION POUR L'EMBEDDING:
  Texte d'entrée: 'full_text'
  Labels: 5 classes
  Exemple de texte: 'Dr. Vikram Prasad is an experienced Dentist in Sowkhya Ayurveda Speciality Clinic, Chennai. He has b...'
 LabelEncoder sauvegardé dans 'label_encoder.pkl'


In [72]:
# Étape 7b: Suppression des mots-métiers des textes (débiaisage)
# Si le texte contient le nom du métier, le classificateur l'apprend trivialement
# → on retire tous les termes de profession pour forcer le modèle à utiliser le style rédactionnel
import re

print("SUPPRESSION DES MOTS-METIERS")
print("=" * 50)

# Construire la liste des termes à supprimer depuis les classes du LabelEncoder
# + variantes plurielles et formes fréquentes
profession_terms = list(le.classes_)

# Variantes supplémentaires courantes dans les bios médicales
extra_terms = [
    "physician", "physicians", "doctor", "doctors",
    "nurse", "nurses", "nursing",
    "surgeon", "surgeons", "surgery",
    "psychiatrist", "psychiatrists", "psychiatry",
    "pediatrician", "pediatricians", "pediatrics", "paediatrics",
    "radiologist", "radiologists", "radiology",
    "cardiologist", "cardiologists", "cardiology",
    "neurologist", "neurologists", "neurology",
    "oncologist", "oncologists", "oncology",
    "dermatologist", "dermatologists", "dermatology",
    "anesthesiologist", "anesthesiologists", "anesthesiology",
    "obstetrician", "obstetricians", "obstetrics",
    "gynecologist", "gynecologists", "gynecology",
    "pathologist", "pathologists", "pathology",
    "internist", "internists",
    "hospitalist", "hospitalists",
    "clinician", "clinicians",
    "practitioner", "practitioners",
    "specialist", "specialists",
    "resident", "residents",
    "fellow", "fellows",
]

# Fusionner et dédupliquer (insensible à la casse)
all_terms = list({t.lower() for t in profession_terms + extra_terms})

# Construire un pattern regex : mot entier, insensible à la casse
pattern = re.compile(
    r'\b(' + '|'.join(re.escape(t) for t in sorted(all_terms, key=len, reverse=True)) + r')\b',
    flags=re.IGNORECASE
)

def remove_profession_words(text):
    return pattern.sub('', text)

# Avant/après sur un exemple
sample = df_train['full_text'].iloc[0]
sample_clean = remove_profession_words(sample)
print(f"Exemple AVANT:\n  {sample[:200]}")
print(f"\nExemple APRES:\n  {sample_clean[:200]}")

# Appliquer sur les 3 splits
df_train['full_text'] = df_train['full_text'].apply(remove_profession_words)
df_val['full_text']   = df_val['full_text'].apply(remove_profession_words)
df_test['full_text']  = df_test['full_text'].apply(remove_profession_words)

print(f"\nTermes supprimes: {len(all_terms)}")
print(f"Traitement applique sur train ({len(df_train)}), val ({len(df_val)}), test ({len(df_test)}) exemples.")
print("Re-lancez les cellules d'extraction d'embeddings (etapes 9-10) pour recalculer avec les textes nettoyés.")


SUPPRESSION DES MOTS-METIERS
Exemple AVANT:
  Dr. Vikram Prasad is an experienced Dentist in Sowkhya Ayurveda Speciality Clinic, Chennai. He has been a practicing Dentist for 20 years. He has done BDS . He is currently associated with Sree Sai De

Exemple APRES:
  Dr. Vikram Prasad is an experienced  in Sowkhya Ayurveda Speciality Clinic, Chennai. He has been a practicing  for 20 years. He has done BDS . He is currently associated with Sree Sai Dental Clinic in

Termes supprimes: 60
Traitement applique sur train (8000), val (1000), test (1000) exemples.
Re-lancez les cellules d'extraction d'embeddings (etapes 9-10) pour recalculer avec les textes nettoyés.


In [1]:
# Étape 8: Configuration de l'embedding avec BERT
print("CONFIGURATION DE L'EMBEDDING BERT")
print("=" * 45)

import subprocess
import sys

pip = [sys.executable, "-m", "pip"]

# Vérifier la version installée de torch via pip (sans importer torch, sans charger les DLL)
result = subprocess.run(pip + ["show", "torch"], capture_output=True, text=True)
torch_version_line = [l for l in result.stdout.splitlines() if l.startswith("Version:")]
torch_version = torch_version_line[0].split(":", 1)[1].strip() if torch_version_line else ""

print(f"Python utilise: {sys.executable}")
print(f"Version torch installee: {torch_version or 'non installee'}")

if "+cu" not in torch_version:
    print("Installation de PyTorch avec support CUDA 12.4...")
    subprocess.run(
        pip + ["install", "--force-reinstall",
               "torch", "torchvision", "torchaudio",
               "--index-url", "https://download.pytorch.org/whl/cu124",
               "--quiet"],
        check=True
    )
    subprocess.run(pip + ["install", "transformers", "--quiet"], check=True)
    print("Installation terminee. Relancez le kernel puis re-executez cette cellule.")
    raise SystemExit("Kernel restart required after PyTorch reinstall.")
else:
    print("PyTorch CUDA deja installe, installation ignoree.")

# Imports pour BERT
from transformers import BertTokenizer, BertModel
import torch

# Vérifier la disponibilité du GPU
if torch.cuda.is_available():
    device = torch.device('cuda')
    print(f"GPU disponible: {torch.cuda.get_device_name(0)}")
    print(f"Memoire GPU: {torch.cuda.get_device_properties(0).total_memory / 1e9:.1f} GB")
else:
    device = torch.device('cpu')
    print("Aucun GPU detecte, utilisation du CPU.")

print(f"Device utilise: {device}")

# Choisir le modèle BERT
model_name = "bert-base-uncased"
print(f"Modele choisi: {model_name}")

# Charger le tokenizer et le modèle BERT
print("Chargement du tokenizer et modele BERT...")
tokenizer = BertTokenizer.from_pretrained(model_name)
model = BertModel.from_pretrained(model_name)
model.to(device)
model.eval()

print("Modele BERT charge!")

# Test rapide du tokenizer sur un exemple
sample_text = df_train['full_text'].iloc[0]
tokens = tokenizer(sample_text, padding=True, truncation=True, max_length=512, return_tensors="pt")

print(f"\nTEST DU TOKENIZER BERT:")
print(f"  Texte original: '{sample_text[:100]}...'")
print(f"  Longueur du texte: {len(sample_text)} caracteres")
print(f"  Nombre de tokens: {tokens['input_ids'].shape[1]}")
print(f"  Max length configuree: 512 tokens")
print(f"  Tokens (premiers 10): {tokenizer.convert_ids_to_tokens(tokens['input_ids'][0][:10])}")

print(f"\nDIMENSIONS D'EMBEDDING:")
print(f"  Vocabulaire size: {tokenizer.vocab_size}")
print(f"  Hidden size: {model.config.hidden_size}")
print(f"  Nombre de layers: {model.config.num_hidden_layers}")
print(f"  Attention heads: {model.config.num_attention_heads}")


CONFIGURATION DE L'EMBEDDING BERT
Python utilise: c:\penv311\Scripts\python.exe
Version torch installee: 2.6.0+cu124
PyTorch CUDA deja installe, installation ignoree.
GPU disponible: NVIDIA GeForce RTX 4050 Laptop GPU
Memoire GPU: 6.4 GB
Device utilise: cuda
Modele choisi: bert-base-uncased
Chargement du tokenizer et modele BERT...


Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: bert-base-uncased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Modele BERT charge!


NameError: name 'df_train' is not defined

In [74]:
# Étape 9: Fonction pour extraire les embeddings BERT
def get_bert_embeddings(texts, batch_size=16):
    """
    Extraire les embeddings BERT pour une liste de textes
    """
    model.eval()
    all_embeddings = []
    
    print(f" Extraction des embeddings pour {len(texts)} textes...")
    
    # Traiter par batch pour éviter les problèmes de mémoire
    for i in range(0, len(texts), batch_size):
        batch_texts = texts[i:i + batch_size]
        
        # Tokenisation du batch
        inputs = tokenizer(
            batch_texts,
            padding=True,
            truncation=True,
            max_length=512,
            return_tensors="pt"
        ).to(device)
        
        # Extraction des embeddings (sans calcul de gradients)
        with torch.no_grad():
            outputs = model(**inputs)
            # Utiliser le [CLS] token (dernière couche, premier token)
            batch_embeddings = outputs.last_hidden_state[:, 0, :].cpu().numpy()
            all_embeddings.append(batch_embeddings)
        
        if (i // batch_size + 1) % 10 == 0:
            print(f"   {i + len(batch_texts)}/{len(texts)} textes traités")
    
    # Concatener tous les embeddings
    final_embeddings = np.concatenate(all_embeddings, axis=0)
    print(f" Extraction terminée! Shape: {final_embeddings.shape}")
    
    return final_embeddings


In [ ]:
# Étape 10: Extraction complète des embeddings pour tous les datasets
import numpy as np

print(" EXTRACTION COMPLÈTE DES EMBEDDINGS BERT")
print("=" * 45)

# Préparer les textes pour chaque dataset
train_texts = df_train['full_text'].tolist()
val_texts = df_val['full_text'].tolist() 
test_texts = df_test['full_text'].tolist()

print(f" DATASETS À TRAITER:")
print(f"  Train: {len(train_texts)} textes")
print(f"  Validation: {len(val_texts)} textes")
print(f"  Test: {len(test_texts)} textes")
print(f"  Total: {len(train_texts) + len(val_texts) + len(test_texts)} textes")

# Extraire les embeddings pour chaque dataset
print(f"\n EXTRACTION EN COURS...")

train_embeddings = get_bert_embeddings(train_texts, batch_size=32)  
val_embeddings = get_bert_embeddings(val_texts, batch_size=32)     
test_embeddings = get_bert_embeddings(test_texts, batch_size=32)

# Préparer les labels
train_labels = df_train['title_encoded'].values
val_labels = df_val['title_encoded'].values 
test_labels = df_test['title_encoded'].values

print(f"\n EXTRACTION TERMINÉE!")
print(f" SHAPES FINALES:")
print(f"  Train: embeddings {train_embeddings.shape}, labels {train_labels.shape}")
print(f"  Valid: embeddings {val_embeddings.shape}, labels {val_labels.shape}")
print(f"  Test: embeddings {test_embeddings.shape}, labels {test_labels.shape}")

# Sauvegarder les embeddings (optionnel, pour éviter de refaire l'extraction)
os.makedirs('embeddings', exist_ok=True)
print(f"\n SAUVEGARDE DES EMBEDDINGS dans embeddings/...")
np.save('embeddings/train_embeddings.npy', train_embeddings)
np.save('embeddings/val_embeddings.npy', val_embeddings) 
np.save('embeddings/test_embeddings.npy', test_embeddings)
np.save('embeddings/train_labels.npy', train_labels)
np.save('embeddings/val_labels.npy', val_labels)
np.save('embeddings/test_labels.npy', test_labels)
print(f" Embeddings sauvegardés dans embeddings/!")


 EXTRACTION COMPLÈTE DES EMBEDDINGS BERT
 DATASETS À TRAITER:
  Train: 8000 textes
  Validation: 1000 textes
  Test: 1000 textes
  Total: 10000 textes

 EXTRACTION EN COURS...
 Extraction des embeddings pour 8000 textes...
   320/8000 textes traités
   640/8000 textes traités
   960/8000 textes traités
   1280/8000 textes traités
   1600/8000 textes traités
   1920/8000 textes traités
   2240/8000 textes traités
   2560/8000 textes traités
   2880/8000 textes traités
   3200/8000 textes traités
   3520/8000 textes traités
   3840/8000 textes traités
   4160/8000 textes traités
   4480/8000 textes traités
   4800/8000 textes traités
   5120/8000 textes traités
   5440/8000 textes traités
   5760/8000 textes traités
   6080/8000 textes traités
   6400/8000 textes traités
   6720/8000 textes traités
   7040/8000 textes traités
   7360/8000 textes traités
   7680/8000 textes traités
   8000/8000 textes traités
 Extraction terminée! Shape: (8000, 768)
 Extraction des embeddings pour 1000 te

In [ ]:
# Étape 11: Création et entraînement du classifieur
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler
import numpy as np
import pickle

print(" CRÉATION ET ENTRAÎNEMENT DU CLASSIFIEUR")
print("=" * 50)

# Charger les embeddings et labels sauvegardés
print(" Chargement des données depuis embeddings/...")
train_embeddings = np.load('embeddings/train_embeddings.npy')
val_embeddings = np.load('embeddings/val_embeddings.npy')
test_embeddings = np.load('embeddings/test_embeddings.npy')
train_labels = np.load('embeddings/train_labels.npy') 
val_labels = np.load('embeddings/val_labels.npy')
test_labels = np.load('embeddings/test_labels.npy')

print(f" Données chargées:")
print(f"  Train: {train_embeddings.shape} -> {train_labels.shape}")
print(f"  Val: {val_embeddings.shape} -> {val_labels.shape}")
print(f"  Test: {test_embeddings.shape} -> {test_labels.shape}")

# Normalisation des embeddings (StandardScaler)
# Fit sur le train uniquement, transform sur val et test
print(f"\n Normalisation des embeddings (StandardScaler)...")
scaler = StandardScaler()
train_embeddings_scaled = scaler.fit_transform(train_embeddings)
val_embeddings_scaled   = scaler.transform(val_embeddings)
test_embeddings_scaled  = scaler.transform(test_embeddings)
print(f"  Moyenne train (avant): {train_embeddings.mean():.4f}  |  après: {train_embeddings_scaled.mean():.4f}")
print(f"  Std train   (avant): {train_embeddings.std():.4f}  |  après: {train_embeddings_scaled.std():.4f}")

# Sauvegarder le scaler pour la comparaison multi-modèles
with open('scaler_bert.pkl', 'wb') as f:
    pickle.dump(scaler, f)
print(f"  Scaler sauvegardé dans 'scaler_bert.pkl'")

# Créer le classifieur Random Forest
print(f"\n Création du classifieur Random Forest...")
rf_classifier = RandomForestClassifier(
    n_estimators=100,
    max_depth=20,
    random_state=42,
    n_jobs=-1
)

# Entraînement
print(f" Entraînement en cours...")
rf_classifier.fit(train_embeddings_scaled, train_labels)
print(f" Entraînement terminé!")

# Prédictions
print(f"\n Évaluation des performances...")
train_pred = rf_classifier.predict(train_embeddings_scaled)
val_pred   = rf_classifier.predict(val_embeddings_scaled)

# Métriques globales
train_acc = accuracy_score(train_labels, train_pred)
val_acc   = accuracy_score(val_labels, val_pred)

print(f" RÉSULTATS:")
print(f"  Accuracy Train:      {train_acc:.4f} ({train_acc*100:.2f}%)")
print(f"  Accuracy Validation: {val_acc:.4f} ({val_acc*100:.2f}%)")


 CRÉATION ET ENTRAÎNEMENT DU CLASSIFIEUR
 Chargement des données...
 Données chargées:
  Train: (8000, 768) -> (8000,)
  Val: (1000, 768) -> (1000,)
  Test: (1000, 768) -> (1000,)

 Normalisation des embeddings (StandardScaler)...
  Moyenne train (avant): -0.0109  |  après: 0.0000
  Std train   (avant): 0.5424  |  après: 1.0000
  Scaler sauvegardé dans 'scaler_bert.pkl'

 Création du classifieur Random Forest...
 Entraînement en cours...
 Entraînement terminé!

 Évaluation des performances...
 RÉSULTATS:
  Accuracy Train:      0.9999 (99.99%)
  Accuracy Validation: 0.7950 (79.50%)


In [77]:
# Résumé des performances
print("=" * 40)
print("RESUME DES PERFORMANCES")
print("=" * 40)
print(f"Modele d'embedding : {model_name}")
print(f"Classifieur        : Random Forest (100 arbres)")
print(f"Accuracy Train     : {train_acc*100:.2f}%")
print(f"Accuracy Val       : {val_acc*100:.2f}%")

# Performances par genre sur la validation
from sklearn.metrics import accuracy_score
df_val_eval = df_val.copy()
df_val_eval['pred'] = val_pred

print("\nPERFORMANCES PAR GENRE (validation):")
for gender in df_val_eval['gender'].unique():
    mask = df_val_eval['gender'] == gender
    acc = accuracy_score(val_labels[mask], val_pred[mask])
    count = mask.sum()
    print(f"  {gender}: {acc*100:.2f}%  (n={count})")


RESUME DES PERFORMANCES
Modele d'embedding : bert-base-uncased
Classifieur        : Random Forest (100 arbres)
Accuracy Train     : 99.99%
Accuracy Val       : 79.50%

PERFORMANCES PAR GENRE (validation):
  Male: 81.65%  (n=436)
  Female: 77.84%  (n=564)


### Comparaison de modèles BERT biomédicaux
Boucle sur 5 modèles : extraction des embeddings, entraînement Random Forest, comparaison de l'accuracy de validation et de l'écart d'accuracy par genre.


In [ ]:
# Étape 12: Boucle de comparaison des modèles biomédicaux
import torch
import numpy as np
import gc
from transformers import AutoTokenizer, AutoModel
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

MODELS_TO_COMPARE = [
    "dmis-lab/biobert-base-cased-v1.2",
    "emilyalsentzer/Bio_ClinicalBERT",
    "allenai/biomed_roberta_base",
    "sultan/BioM-ELECTRA-Large-SQuAD2",
    "microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract",
]

def extract_cls_embeddings(mdl, tok, texts, dev, batch_size=16):
    mdl.eval()
    all_emb = []
    for i in range(0, len(texts), batch_size):
        batch = texts[i:i + batch_size]
        inputs = tok(batch, padding=True, truncation=True, max_length=512, return_tensors="pt").to(dev)
        with torch.no_grad():
            outputs = mdl(**inputs)
            emb = outputs.last_hidden_state[:, 0, :].cpu().numpy()
        all_emb.append(emb)
        if (i // batch_size + 1) % 20 == 0:
            print(f"    {i + len(batch)}/{len(texts)} traites")
    return np.concatenate(all_emb, axis=0)

# Textes et labels (déjà chargés)
train_texts_cmp = df_train['full_text'].tolist()
val_texts_cmp   = df_val['full_text'].tolist()
train_lbl_cmp   = df_train['title_encoded'].values
val_lbl_cmp     = df_val['title_encoded'].values

comparison_results = []

for model_id in MODELS_TO_COMPARE:
    print(f"\n{'='*60}")
    print(f"MODELE: {model_id}")
    print(f"{'='*60}")
    bs = 8 if "large" in model_id.lower() else 16
    try:
        print("  Chargement du tokenizer et du modele...")
        tok = AutoTokenizer.from_pretrained(model_id)
        mdl = AutoModel.from_pretrained(model_id)
        mdl.to(device)
        mdl.eval()

        print(f"  Extraction embeddings train ({len(train_texts_cmp)} textes, batch={bs})...")
        tr_emb = extract_cls_embeddings(mdl, tok, train_texts_cmp, device, bs)
        print(f"  Extraction embeddings val ({len(val_texts_cmp)} textes)...")
        va_emb = extract_cls_embeddings(mdl, tok, val_texts_cmp, device, bs)

        # Normalisation des embeddings
        print("  Normalisation des embeddings (StandardScaler)...")
        sc = StandardScaler()
        tr_emb_scaled = sc.fit_transform(tr_emb)
        va_emb_scaled = sc.transform(va_emb)

        # Sauvegarde des embeddings normalisés dans embeddings/
        os.makedirs('embeddings', exist_ok=True)
        safe_name = model_id.replace("/", "_")
        np.save(f'embeddings/train_emb_{safe_name}.npy', tr_emb_scaled)
        np.save(f'embeddings/val_emb_{safe_name}.npy', va_emb_scaled)

        # Entraînement Random Forest
        print("  Entrainement Random Forest...")
        clf = RandomForestClassifier(n_estimators=100, max_depth=20, random_state=42, n_jobs=-1)
        clf.fit(tr_emb_scaled, train_lbl_cmp)

        tr_pred = clf.predict(tr_emb_scaled)
        va_pred = clf.predict(va_emb_scaled)
        tr_acc  = accuracy_score(train_lbl_cmp, tr_pred)
        va_acc  = accuracy_score(val_lbl_cmp, va_pred)

        # Accuracy par genre
        gender_accs = {}
        for g in df_val['gender'].unique():
            mask_g = df_val['gender'] == g
            gender_accs[g] = accuracy_score(val_lbl_cmp[mask_g], va_pred[mask_g])

        comparison_results.append({
            "model": model_id,
            "train_acc": tr_acc,
            "val_acc": va_acc,
            "gender_accs": gender_accs,
            "status": "OK"
        })
        print(f"  Train Acc: {tr_acc*100:.2f}%  |  Val Acc: {va_acc*100:.2f}%")
        for g, a in gender_accs.items():
            print(f"    {g}: {a*100:.2f}%")

    except Exception as e:
        print(f"  ERREUR: {e}")
        comparison_results.append({
            "model": model_id, "train_acc": None,
            "val_acc": None, "gender_accs": {}, "status": str(e)
        })

    finally:
        for var in ['mdl', 'tok', 'tr_emb', 'va_emb', 'tr_emb_scaled', 'va_emb_scaled', 'sc', 'clf']:
            try: del locals()[var]
            except: pass
        gc.collect()
        if torch.cuda.is_available():
            torch.cuda.empty_cache()

print("\nComparaison terminee!")



MODELE: dmis-lab/biobert-base-cased-v1.2
  Chargement du tokenizer et du modele...


Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: dmis-lab/biobert-base-cased-v1.2
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.decoder.bias               | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.decoder.weight             | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


  Extraction embeddings train (8000 textes, batch=16)...
    320/8000 traites
    640/8000 traites
    960/8000 traites
    1280/8000 traites
    1600/8000 traites
    1920/8000 traites
    2240/8000 traites
    2560/8000 traites
    2880/8000 traites
    3200/8000 traites
    3520/8000 traites
    3840/8000 traites
    4160/8000 traites
    4480/8000 traites
    4800/8000 traites
    5120/8000 traites
    5440/8000 traites
    5760/8000 traites
    6080/8000 traites
    6400/8000 traites
    6720/8000 traites
    7040/8000 traites
    7360/8000 traites
    7680/8000 traites
    8000/8000 traites
  Extraction embeddings val (1000 textes)...
    320/1000 traites
    640/1000 traites
    960/1000 traites
  Normalisation des embeddings (StandardScaler)...
  Entrainement Random Forest...
  Train Acc: 100.00%  |  Val Acc: 71.40%
    Male: 75.23%
    Female: 68.44%

MODELE: emilyalsentzer/Bio_ClinicalBERT
  Chargement du tokenizer et du modele...


Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: emilyalsentzer/Bio_ClinicalBERT
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.decoder.weight             | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


  Extraction embeddings train (8000 textes, batch=16)...
    320/8000 traites
    640/8000 traites
    960/8000 traites
    1280/8000 traites
    1600/8000 traites
    1920/8000 traites
    2240/8000 traites
    2560/8000 traites
    2880/8000 traites
    3200/8000 traites
    3520/8000 traites
    3840/8000 traites
    4160/8000 traites
    4480/8000 traites
    4800/8000 traites
    5120/8000 traites
    5440/8000 traites
    5760/8000 traites
    6080/8000 traites
    6400/8000 traites
    6720/8000 traites
    7040/8000 traites
    7360/8000 traites
    7680/8000 traites
    8000/8000 traites
  Extraction embeddings val (1000 textes)...
    320/1000 traites
    640/1000 traites
    960/1000 traites
  Normalisation des embeddings (StandardScaler)...
  Entrainement Random Forest...
  Train Acc: 99.99%  |  Val Acc: 72.50%
    Male: 75.69%
    Female: 70.04%

MODELE: allenai/biomed_roberta_base
  Chargement du tokenizer et du modele...


Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

RobertaModel LOAD REPORT from: allenai/biomed_roberta_base
Key                       | Status     |  | 
--------------------------+------------+--+-
lm_head.bias              | UNEXPECTED |  | 
lm_head.layer_norm.weight | UNEXPECTED |  | 
lm_head.layer_norm.bias   | UNEXPECTED |  | 
lm_head.dense.bias        | UNEXPECTED |  | 
lm_head.dense.weight      | UNEXPECTED |  | 
lm_head.decoder.weight    | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


  Extraction embeddings train (8000 textes, batch=16)...
    320/8000 traites
    640/8000 traites
    960/8000 traites
    1280/8000 traites
    1600/8000 traites
    1920/8000 traites
    2240/8000 traites
    2560/8000 traites
    2880/8000 traites
    3200/8000 traites
    3520/8000 traites
    3840/8000 traites
    4160/8000 traites
    4480/8000 traites
    4800/8000 traites
    5120/8000 traites
    5440/8000 traites
    5760/8000 traites
    6080/8000 traites
    6400/8000 traites
    6720/8000 traites
    7040/8000 traites
    7360/8000 traites
    7680/8000 traites
    8000/8000 traites
  Extraction embeddings val (1000 textes)...
    320/1000 traites
    640/1000 traites
    960/1000 traites
  Normalisation des embeddings (StandardScaler)...
  Entrainement Random Forest...
  Train Acc: 99.89%  |  Val Acc: 83.70%
    Male: 84.17%
    Female: 83.33%

MODELE: sultan/BioM-ELECTRA-Large-SQuAD2
  Chargement du tokenizer et du modele...


Loading weights:   0%|          | 0/389 [00:00<?, ?it/s]

ElectraModel LOAD REPORT from: sultan/BioM-ELECTRA-Large-SQuAD2
Key                             | Status     |  | 
--------------------------------+------------+--+-
qa_outputs.bias                 | UNEXPECTED |  | 
electra.embeddings.position_ids | UNEXPECTED |  | 
qa_outputs.weight               | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


  Extraction embeddings train (8000 textes, batch=8)...
    160/8000 traites
    320/8000 traites
    480/8000 traites
    640/8000 traites
    800/8000 traites
    960/8000 traites
    1120/8000 traites
    1280/8000 traites
    1440/8000 traites
    1600/8000 traites
    1760/8000 traites
    1920/8000 traites
    2080/8000 traites
    2240/8000 traites
    2400/8000 traites
    2560/8000 traites
    2720/8000 traites
    2880/8000 traites
    3040/8000 traites
    3200/8000 traites
    3360/8000 traites
    3520/8000 traites
    3680/8000 traites
    3840/8000 traites
    4000/8000 traites
    4160/8000 traites
    4320/8000 traites
    4480/8000 traites
    4640/8000 traites
    4800/8000 traites
    4960/8000 traites
    5120/8000 traites
    5280/8000 traites
    5440/8000 traites
    5600/8000 traites
    5760/8000 traites
    5920/8000 traites
    6080/8000 traites
    6240/8000 traites
    6400/8000 traites
    6560/8000 traites
    6720/8000 traites
    6880/8000 traites
    

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.decoder.bias               | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.decoder.weight             | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


  Extraction embeddings train (8000 textes, batch=16)...
    320/8000 traites
    640/8000 traites
    960/8000 traites
    1280/8000 traites
    1600/8000 traites
    1920/8000 traites
    2240/8000 traites
    2560/8000 traites
    2880/8000 traites
    3200/8000 traites
    3520/8000 traites
    3840/8000 traites
    4160/8000 traites
    4480/8000 traites
    4800/8000 traites
    5120/8000 traites
    5440/8000 traites
    5760/8000 traites
    6080/8000 traites
    6400/8000 traites
    6720/8000 traites
    7040/8000 traites
    7360/8000 traites
    7680/8000 traites
    8000/8000 traites
  Extraction embeddings val (1000 textes)...
    320/1000 traites
    640/1000 traites
    960/1000 traites
  Normalisation des embeddings (StandardScaler)...
  Entrainement Random Forest...
  Train Acc: 100.00%  |  Val Acc: 80.70%
    Male: 80.73%
    Female: 80.67%

Comparaison terminee!


In [79]:
# Étape 13: Tableau comparatif des modèles
import pandas as pd

print("=" * 75)
print("TABLEAU COMPARATIF DES MODELES BIOMEDICAUX")
print("=" * 75)

rows = []
for r in comparison_results:
    short_name = r["model"].split("/")[-1]
    row = {
        "Modele": short_name,
        "Train Acc": f"{r['train_acc']*100:.2f}%" if r["train_acc"] is not None else "ERREUR",
        "Val Acc":   f"{r['val_acc']*100:.2f}%"   if r["val_acc"]   is not None else "ERREUR",
    }
    for g, a in r.get("gender_accs", {}).items():
        row[f"Val ({g})"] = f"{a*100:.2f}%"
    rows.append(row)

df_compare = pd.DataFrame(rows)
print(df_compare.to_string(index=False))

# Meilleur modèle selon Val Acc
valid = [r for r in comparison_results if r["val_acc"] is not None]
if valid:
    best = max(valid, key=lambda x: x["val_acc"])
    print(f"\nMEILLEUR MODELE  : {best['model']}")
    print(f"Val Accuracy     : {best['val_acc']*100:.2f}%")
    # Écart d'accuracy entre genres (fairness gap)
    if len(best["gender_accs"]) >= 2:
        accs = list(best["gender_accs"].values())
        gap = max(accs) - min(accs)
        print(f"Ecart genre (gap): {gap*100:.2f}%")


TABLEAU COMPARATIF DES MODELES BIOMEDICAUX
                                    Modele Train Acc Val Acc Val (Male) Val (Female)
                   biobert-base-cased-v1.2   100.00%  71.40%     75.23%       68.44%
                          Bio_ClinicalBERT    99.99%  72.50%     75.69%       70.04%
                       biomed_roberta_base    99.89%  83.70%     84.17%       83.33%
                 BioM-ELECTRA-Large-SQuAD2   100.00%  63.40%     65.83%       61.52%
BiomedNLP-BiomedBERT-base-uncased-abstract   100.00%  80.70%     80.73%       80.67%

MEILLEUR MODELE  : allenai/biomed_roberta_base
Val Accuracy     : 83.70%
Ecart genre (gap): 0.84%


In [81]:
# Cellule de collecte des résultats pour le README
from sklearn.metrics import accuracy_score

print("=== RESULTATS BERT-BASE-UNCASED (SANS suppression mots-metiers) ===")
print(f"train_acc={train_acc*100:.2f}")
print(f"val_acc={val_acc*100:.2f}")
df_val_tmp = df_val.copy()
df_val_tmp['pred'] = val_pred
for g in sorted(df_val_tmp['gender'].unique()):
    mask = df_val_tmp['gender'] == g
    a = accuracy_score(val_labels[mask], val_pred[mask])
    print(f"val_acc_{g}={a*100:.2f} n={mask.sum()}")

print("\n=== COMPARAISON MODELES BIOMEDICAUX ===")
for r in comparison_results:
    gaps = list(r['gender_accs'].values())
    gap = max(gaps) - min(gaps) if len(gaps) >= 2 else 0
    print(f"model={r['model'].split('/')[-1]} | train={r['train_acc']*100:.2f} | val={r['val_acc']*100:.2f} | gap={gap*100:.2f} | " +
          " | ".join(f"{g}={a*100:.2f}" for g, a in r['gender_accs'].items()))


=== RESULTATS BERT-BASE-UNCASED (SANS suppression mots-metiers) ===
train_acc=99.99
val_acc=79.50
val_acc_Female=77.84 n=564
val_acc_Male=81.65 n=436

=== COMPARAISON MODELES BIOMEDICAUX ===
model=biobert-base-cased-v1.2 | train=100.00 | val=71.40 | gap=6.79 | Male=75.23 | Female=68.44
model=Bio_ClinicalBERT | train=99.99 | val=72.50 | gap=5.65 | Male=75.69 | Female=70.04
model=biomed_roberta_base | train=99.89 | val=83.70 | gap=0.84 | Male=84.17 | Female=83.33
model=BioM-ELECTRA-Large-SQuAD2 | train=100.00 | val=63.40 | gap=4.30 | Male=65.83 | Female=61.52
model=BiomedNLP-BiomedBERT-base-uncased-abstract | train=100.00 | val=80.70 | gap=0.06 | Male=80.73 | Female=80.67


---
## Restructuration : D0 / D1 / D1n

| Version | Description |
|---------|-------------|
| **D0** | Texte brut, aucun nettoyage |
| **D1** | Mots-métiers masqués (débiaisage lexical) |
| **D1n** | D1 + neutralisation des indices de genre (pronoms, titres, relations familiales) |


In [3]:
# === Création de D0, D1, D1n ===
import json
from sklearn.preprocessing import LabelEncoder
import pickle
# === Imports globaux ===
import os
import re
import gc
import subprocess
import sys

import numpy as np
import pandas as pd
import torch

from transformers import AutoTokenizer, AutoModel
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix, classification_report

print("Imports OK.")


def load_jsonl(path):
    data = []
    with open(path, 'r', encoding='utf-8') as f:
        for line in f:
            if line.strip():
                data.append(json.loads(line))
    return data

# Rechargement propre des données brutes
df0_train = pd.DataFrame(load_jsonl('medical_bios_data/train.jsonl'))
df0_val   = pd.DataFrame(load_jsonl('medical_bios_data/validation.jsonl'))
df0_test  = pd.DataFrame(load_jsonl('medical_bios_data/test.jsonl'))

# LabelEncoder fit sur train uniquement (correction)
le_d = LabelEncoder()
le_d.fit(df0_train['title'])
for df in [df0_train, df0_val, df0_test]:
    df['label'] = le_d.transform(df['title'])

# ── D0 : texte brut ──────────────────────────────────────────────
d0_train = df0_train[['full_text', 'gender', 'label']].copy()
d0_val   = df0_val[['full_text', 'gender', 'label']].copy()
d0_test  = df0_test[['full_text', 'gender', 'label']].copy()

# ── D1 : masquage des mots-métiers ───────────────────────────────
profession_terms = list(le_d.classes_)
extra_terms = [
    "physician","physicians","doctor","doctors","nurse","nurses","nursing",
    "surgeon","surgeons","surgery","psychiatrist","psychiatrists","psychiatry",
    "pediatrician","pediatricians","pediatrics","paediatrics",
    "radiologist","radiologists","radiology","cardiologist","cardiologists","cardiology",
    "neurologist","neurologists","neurology","oncologist","oncologists","oncology",
    "dermatologist","dermatologists","dermatology",
    "anesthesiologist","anesthesiologists","anesthesiology",
    "obstetrician","obstetricians","obstetrics","gynecologist","gynecologists","gynecology",
    "pathologist","pathologists","pathology","internist","internists",
    "hospitalist","hospitalists","clinician","clinicians",
    "practitioner","practitioners","specialist","specialists","resident","residents","fellow","fellows",
]
all_prof_terms = list({t.lower() for t in profession_terms + extra_terms})
prof_pattern = re.compile(
    r'\b(' + '|'.join(re.escape(t) for t in sorted(all_prof_terms, key=len, reverse=True)) + r')\b',
    flags=re.IGNORECASE
)

def mask_professions(text):
    return prof_pattern.sub('[PROF]', text)

d1_train = df0_train[['full_text', 'gender', 'label']].copy()
d1_val   = df0_val[['full_text', 'gender', 'label']].copy()
d1_test  = df0_test[['full_text', 'gender', 'label']].copy()
for df in [d1_train, d1_val, d1_test]:
    df['full_text'] = df['full_text'].apply(mask_professions)

# ── D1n : D1 + neutralisation des indices de genre ───────────────
GENDER_MAP = {
    'he': '[PRONOUN]', 'she': '[PRONOUN]',
    'him': '[PRONOUN]', 'his': '[POSS]', 'her': '[POSS]',
    'himself': '[PRONOUN]', 'herself': '[PRONOUN]',
    'mr': '[TITLE]', 'mr.': '[TITLE]', 'mrs': '[TITLE]', 'mrs.': '[TITLE]',
    'ms': '[TITLE]', 'ms.': '[TITLE]', 'miss': '[TITLE]', 'dr': '[TITLE]',
    'mother': '[PARENT]', 'father': '[PARENT]',
    'wife': '[SPOUSE]', 'husband': '[SPOUSE]',
    'son': '[CHILD]', 'daughter': '[CHILD]',
    'brother': '[SIBLING]', 'sister': '[SIBLING]',
    'grandfather': '[GRANDPARENT]', 'grandmother': '[GRANDPARENT]',
    'nephew': '[RELATIVE]', 'niece': '[RELATIVE]',
    'boyfriend': '[PARTNER]', 'girlfriend': '[PARTNER]',
}
gender_pattern = re.compile(
    r'\b(' + '|'.join(re.escape(k) for k in sorted(GENDER_MAP, key=len, reverse=True)) + r')\.?\b',
    flags=re.IGNORECASE
)

def neutralize_gender(text):
    def _replace(m):
        key = m.group(0).lower().rstrip('.')
        return GENDER_MAP.get(key, m.group(0))
    return gender_pattern.sub(_replace, text)

d1n_train = d1_train.copy()
d1n_val   = d1_val.copy()
d1n_test  = d1_test.copy()
for df in [d1n_train, d1n_val, d1n_test]:
    df['full_text'] = df['full_text'].apply(neutralize_gender)

print("Versions créées :")
print(f"  D0  – brut         : {len(d0_train)} train, {len(d0_val)} val, {len(d0_test)} test")
print(f"  D1  – prof masquée : {len(d1_train)} train, {len(d1_val)} val, {len(d1_test)} test")
print(f"  D1n – prof+genre   : {len(d1n_train)} train, {len(d1n_val)} val, {len(d1n_test)} test")

# Exemple
print("\n── Exemple D0 ──")
print(d0_train['full_text'].iloc[0][:200])
print("\n── Exemple D1 ──")
print(d1_train['full_text'].iloc[0][:200])
print("\n── Exemple D1n ──")
print(d1n_train['full_text'].iloc[0][:200])

# Sauvegarder le LabelEncoder corrigé
with open('label_encoder_d.pkl', 'wb') as f:
    pickle.dump(le_d, f)
print(f"\nLabelEncoder (fit sur train uniquement) sauvegardé.")


Imports OK.
Versions créées :
  D0  – brut         : 8000 train, 1000 val, 1000 test
  D1  – prof masquée : 8000 train, 1000 val, 1000 test
  D1n – prof+genre   : 8000 train, 1000 val, 1000 test

── Exemple D0 ──
Dr. Vikram Prasad is an experienced Dentist in Sowkhya Ayurveda Speciality Clinic, Chennai. He has been a practicing Dentist for 20 years. He has done BDS . He is currently associated with Sree Sai De

── Exemple D1 ──
Dr. Vikram Prasad is an experienced [PROF] in Sowkhya Ayurveda Speciality Clinic, Chennai. He has been a practicing [PROF] for 20 years. He has done BDS . He is currently associated with Sree Sai Dent

── Exemple D1n ──
[TITLE]. Vikram Prasad is an experienced [PROF] in Sowkhya Ayurveda Speciality Clinic, Chennai. [PRONOUN] has been a practicing [PROF] for 20 years. [PRONOUN] has done BDS . [PRONOUN] is currently ass

LabelEncoder (fit sur train uniquement) sauvegardé.


### Analyse du label leakage

Quantification du nombre de termes de profession présents dans les textes bruts (D0).


In [9]:
# === Analyse du label leakage ===
import pandas as pd
import numpy as np

def count_prof_matches(text):
    return len(prof_pattern.findall(text))

# Sur le train D0
leakage = []
for _, row in d0_train.iterrows():
    matches = prof_pattern.findall(row['full_text'])
    leakage.append({'profession': row['label'], 'n_matches': len(matches), 'has_leak': len(matches) > 0})

df_leak = pd.DataFrame(leakage)
n_total  = len(df_leak)
n_leaked = df_leak['has_leak'].sum()
pct      = n_leaked / n_total * 100

print("=" * 55)
print("ANALYSE DU LABEL LEAKAGE (D0 – train)")
print("=" * 55)
print(f"  Textes avec ≥1 terme de profession : {n_leaked}/{n_total} ({pct:.1f}%)")
print(f"  Occurrences totales masquées         : {df_leak['n_matches'].sum()}")
print(f"  Moyenne d'occurrences par texte      : {df_leak['n_matches'].mean():.2f}")
print(f"  Max occurrences dans un texte        : {df_leak['n_matches'].max()}")

# Par profession
print("\n  Taux de leakage par profession:")
prof_names = dict(zip(range(len(le_d.classes_)), le_d.classes_))
leak_by_prof = df_leak.groupby('profession')['has_leak'].mean().sort_values(ascending=False)
for code, rate in leak_by_prof.items():
    name = prof_names.get(code, str(code))
    print(f"    {name:35s}: {rate*100:.1f}%")

# Exemples avant/après masquage
print("\n  ── Exemples avant/après masquage ──")
examples = d0_train[d0_train['full_text'].apply(count_prof_matches) > 0].head(3)
for i, (_, row) in enumerate(examples.iterrows()):
    raw  = row['full_text'][:250]
    prof_name = le_d.classes_[row['label']]
    masked = prof_pattern.sub('[PROF]', raw)
    print(f"\n  Exemple {i+1} ({prof_name}):")
    print(f"  AVANT  : {raw}")
    print(f"  APRES  : {masked}")


ANALYSE DU LABEL LEAKAGE (D0 – train)
  Textes avec ≥1 terme de profession : 7639/8000 (95.5%)
  Occurrences totales masquées         : 19189
  Moyenne d'occurrences par texte      : 2.40
  Max occurrences dans un texte        : 16

  Taux de leakage par profession:
    dentist                            : 100.0%
    nurse                              : 100.0%
    physician                          : 100.0%
    surgeon                            : 96.4%
    psychologist                       : 85.7%

  ── Exemples avant/après masquage ──

  Exemple 1 (dentist):
  AVANT  : Dr. Vikram Prasad is an experienced Dentist in Sowkhya Ayurveda Speciality Clinic, Chennai. He has been a practicing Dentist for 20 years. He has done BDS . He is currently associated with Sree Sai Dental Clinic in Sowkhya Ayurveda Speciality Clinic,
  APRES  : Dr. Vikram Prasad is an experienced [PROF] in Sowkhya Ayurveda Speciality Clinic, Chennai. He has been a practicing [PROF] for 20 years. He has done BDS . He i

---
## Pipeline : embeddings + Random Forest sur D0 / D1 / D1n

Extraction des embeddings BiomedBERT (frozen) sur les trois versions de données, entraînement Random Forest, et évaluation avec les métriques de fairness complètes.


In [4]:
# === Fonctions utilitaires : métriques fairness ===

def compute_fairness_metrics(y_true, y_pred, gender_series, classes):
    """
    Calcule, pour chaque groupe de genre, les métriques suivantes :
      - accuracy, macro-F1, weighted-F1
      - TPR moyen (Equal Opportunity)
      - FPR moyen
      - Demographic Parity (taux de prédiction positive moyen)
    Retourne aussi les gaps inter-groupes.
    """
    groups = sorted(gender_series.unique())
    group_stats = {}

    for g in groups:
        mask = (gender_series == g).values
        yt = np.array(y_true)[mask]
        yp = np.array(y_pred)[mask]

        acc  = accuracy_score(yt, yp)
        mf1  = f1_score(yt, yp, average='macro',    zero_division=0)
        wf1  = f1_score(yt, yp, average='weighted', zero_division=0)

        tprs, fprs, pred_rates = [], [], []
        for c in classes:
            yt_b = (yt == c).astype(int)
            yp_b = (yp == c).astype(int)
            if yt_b.sum() == 0:
                continue
            tn, fp, fn, tp = confusion_matrix(yt_b, yp_b, labels=[0, 1]).ravel()
            tprs.append(tp / (tp + fn) if (tp + fn) > 0 else 0.0)
            fprs.append(fp / (fp + tn) if (fp + tn) > 0 else 0.0)
            pred_rates.append(yp_b.mean())

        group_stats[g] = {
            'n': int(mask.sum()),
            'accuracy': acc,
            'macro_f1': mf1,
            'weighted_f1': wf1,
            'mean_tpr': float(np.mean(tprs)) if tprs else 0.0,
            'mean_fpr': float(np.mean(fprs)) if fprs else 0.0,
            'mean_pred_rate': float(np.mean(pred_rates)) if pred_rates else 0.0,
        }

    accs  = [group_stats[g]['accuracy']       for g in groups]
    tprs  = [group_stats[g]['mean_tpr']       for g in groups]
    fprs  = [group_stats[g]['mean_fpr']       for g in groups]
    prs   = [group_stats[g]['mean_pred_rate'] for g in groups]
    mf1s  = [group_stats[g]['macro_f1']       for g in groups]

    return {
        'groups': group_stats,
        'dp_gap':   max(prs)  - min(prs),    # Demographic Parity gap
        'eopp_gap': max(tprs) - min(tprs),   # Equal Opportunity / TPR gap
        'fpr_gap':  max(fprs) - min(fprs),   # FPR gap
        'acc_gap':  max(accs) - min(accs),   # accuracy gap (simple)
        'f1_gap':   max(mf1s) - min(mf1s),   # macro-F1 gap
    }


def print_fairness_report(label, fairness_dict, y_true, y_pred, classes, le):
    """Affiche un rapport lisible des métriques fairness."""
    print(f"\n{'='*60}")
    print(f"FAIRNESS REPORT — {label}")
    print(f"{'='*60}")
    print(f"  Accuracy globale : {accuracy_score(y_true, y_pred)*100:.2f}%")
    print(f"  Macro-F1 global  : {f1_score(y_true, y_pred, average='macro', zero_division=0)*100:.2f}%")
    print(f"  Weighted-F1      : {f1_score(y_true, y_pred, average='weighted', zero_division=0)*100:.2f}%")

    print(f"\n  ── Par genre ──")
    for g, s in fairness_dict['groups'].items():
        print(f"  {g:8s}: acc={s['accuracy']*100:.2f}%  macro-F1={s['macro_f1']*100:.2f}%  "
              f"TPR={s['mean_tpr']*100:.2f}%  FPR={s['mean_fpr']*100:.2f}%  (n={s['n']})")

    print(f"\n  ── Gaps inter-groupes ──")
    print(f"  Accuracy gap (acc_gap)           : {fairness_dict['acc_gap']*100:.2f}%")
    print(f"  Demographic Parity gap (dp_gap)  : {fairness_dict['dp_gap']*100:.2f}%")
    print(f"  Equal Opportunity gap (eopp_gap) : {fairness_dict['eopp_gap']*100:.2f}%")
    print(f"  FPR gap                          : {fairness_dict['fpr_gap']*100:.2f}%")
    print(f"  Macro-F1 gap                     : {fairness_dict['f1_gap']*100:.2f}%")

print("Fonctions de métriques fairness définies.")


Fonctions de métriques fairness définies.


In [5]:
# === Pipeline : extraction embeddings + RF + fairness sur D0/D1/D1n ===


MAIN_MODEL_ID = "microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract"

def cls_embeddings(model, tokenizer, texts, device, batch_size=16):
    model.eval()
    all_emb = []
    for i in range(0, len(texts), batch_size):
        batch = texts[i:i + batch_size]
        inputs = tokenizer(batch, padding=True, truncation=True,
                           max_length=512, return_tensors="pt").to(device)
        with torch.no_grad():
            out = model(**inputs)
            all_emb.append(out.last_hidden_state[:, 0, :].cpu().numpy())
        if (i // batch_size + 1) % 30 == 0:
            print(f"    {i+len(batch)}/{len(texts)}")
    return np.concatenate(all_emb, axis=0)

# Chargement du modèle principal une seule fois
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Device : {device}")
print(f"Chargement de {MAIN_MODEL_ID}...")
tok_main = AutoTokenizer.from_pretrained(MAIN_MODEL_ID)
mdl_main = AutoModel.from_pretrained(MAIN_MODEL_ID).to(device)

classes = list(range(len(le_d.classes_)))
fairness_results = {}  # {version_label: {fairness_dict, y_true, y_pred, gender}}

DATASETS = {
    "D0": (d0_train, d0_val),
    "D1": (d1_train, d1_val),
    "D1n": (d1n_train, d1n_val),
}

for version, (tr_df, va_df) in DATASETS.items():
    print(f"\n{'─'*50}")
    print(f"VERSION {version}")
    print(f"{'─'*50}")

    tr_texts = tr_df['full_text'].tolist()
    va_texts = va_df['full_text'].tolist()
    tr_lbl   = tr_df['label'].values
    va_lbl   = va_df['label'].values
    va_gender = va_df['gender']

    print(f"  Extraction embeddings train ({len(tr_texts)})...")
    tr_emb = cls_embeddings(mdl_main, tok_main, tr_texts, device, batch_size=16)
    print(f"  Extraction embeddings val ({len(va_texts)})...")
    va_emb = cls_embeddings(mdl_main, tok_main, va_texts, device, batch_size=16)

    sc = StandardScaler()
    tr_emb_s = sc.fit_transform(tr_emb)
    va_emb_s = sc.transform(va_emb)

    # Sauvegarder
    os.makedirs('embeddings', exist_ok=True)
    np.save(f'embeddings/train_emb_{version}.npy', tr_emb_s)
    np.save(f'embeddings/val_emb_{version}.npy',   va_emb_s)

    print("  Entraînement Random Forest...")
    clf = RandomForestClassifier(n_estimators=100, max_depth=20, random_state=42, n_jobs=-1)
    clf.fit(tr_emb_s, tr_lbl)
    va_pred = clf.predict(va_emb_s)

    fd = compute_fairness_metrics(va_lbl, va_pred, va_gender, classes)
    fairness_results[version] = {
        'fairness': fd,
        'y_true': va_lbl,
        'y_pred': va_pred,
        'gender': va_gender,
    }
    print_fairness_report(f"BiomedBERT frozen + RF — {version}", fd, va_lbl, va_pred, classes, le_d)

print("\n\nExtraction terminée pour toutes les versions.")


Device : cuda
Chargement de microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract...


Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.predictions.decoder.bias               | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.predictions.decoder.weight             | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.



──────────────────────────────────────────────────
VERSION D0
──────────────────────────────────────────────────
  Extraction embeddings train (8000)...
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
  Extraction embeddings val (1000)...
    480/1000
    960/1000
  Entraînement Random Forest...

FAIRNESS REPORT — BiomedBERT frozen + RF — D0
  Accuracy globale : 87.20%
  Macro-F1 global  : 85.95%
  Weighted-F1      : 86.88%

  ── Par genre ──
  Female  : acc=85.99%  macro-F1=80.14%  TPR=78.48%  FPR=4.01%  (n=564)
  Male    : acc=88.76%  macro-F1=87.33%  TPR=87.28%  FPR=2.96%  (n=436)

  ── Gaps inter-groupes ──
  Accuracy gap (acc_gap)           : 2.77%
  Demographic Parity gap (dp_gap)  : 0.00%
  Equal Opportunity gap (eopp_gap) : 8.80%
  FPR gap                          : 1.05%
  Macro-F1 gap                    

---
## Robustesse — Gender Swap & Flip Rate

Pour chaque texte de test, on permute les pronoms de genre (`he↔she`, `his↔her`, `him↔her`, `himself↔herself`), on ré-extrait les embeddings et on mesure le **flip rate** = proportion de prédictions qui changent.

Un flip rate élevé indique que le modèle s'appuie sur les indices de genre plutôt que sur le contenu médical.


In [13]:
# === Gender Swap + Flip Rate ===
# Dictionnaire de permutation (préserve la casse)
SWAP_PAIRS = [
    ('he', 'she'), ('she', 'he'),
    ('him', 'her'), ('her', 'him'),
    ('his', 'her'), ('himself', 'herself'), ('herself', 'himself'),
    ('He', 'She'), ('She', 'He'),
    ('Him', 'Her'), ('Her', 'Him'),
    ('His', 'Her'), ('Himself', 'Herself'), ('Herself', 'Himself'),
    ('HE', 'SHE'), ('SHE', 'HE'),
]
# Construire la regex dans l'ordre décroissant de longueur pour éviter les collisions
_swap_dict = {k: v for k, v in SWAP_PAIRS}
_swap_pattern = re.compile(
    r'\b(' + '|'.join(re.escape(k) for k in sorted(_swap_dict, key=len, reverse=True)) + r')\b'
)

def gender_swap(text):
    return _swap_pattern.sub(lambda m: _swap_dict.get(m.group(0), m.group(0)), text)

# Appliqué sur D0 val (texte brut : le + grand signal de genre)
val_texts_d0 = d0_val['full_text'].tolist()
val_lbl_gs   = d0_val['label'].values
val_gender_gs = d0_val['gender']

val_texts_swapped = [gender_swap(t) for t in val_texts_d0]

# Vérification sur un exemple
idx_check = next((i for i, t in enumerate(val_texts_d0) if ' he ' in t.lower() or ' she ' in t.lower()), 0)
print("── Exemple gender swap ──")
print("ORIGINAL :", val_texts_d0[idx_check][:200])
print("SWAPPED  :", val_texts_swapped[idx_check][:200])

# Extraction des embeddings originaux et swappés
print(f"\nExtraction embeddings originaux ({len(val_texts_d0)} textes)...")
emb_orig   = cls_embeddings(mdl_main, tok_main, val_texts_d0,    device, batch_size=16)
print(f"Extraction embeddings swappés  ({len(val_texts_swapped)} textes)...")
emb_swapped = cls_embeddings(mdl_main, tok_main, val_texts_swapped, device, batch_size=16)

# Normalisation avec le scaler D0 (rechargé ou refitté)
sc_gs = StandardScaler()
emb_orig_s   = sc_gs.fit_transform(emb_orig)
emb_swapped_s = sc_gs.transform(emb_swapped)

# Entraîner RF sur D0 train
tr_texts_d0 = d0_train['full_text'].tolist()
tr_lbl_d0   = d0_train['label'].values
emb_tr_d0   = cls_embeddings(mdl_main, tok_main, tr_texts_d0, device, batch_size=16)
emb_tr_d0_s = sc_gs.fit_transform(emb_tr_d0)
emb_orig_s  = sc_gs.transform(emb_orig)
emb_swapped_s = sc_gs.transform(emb_swapped)

clf_gs = RandomForestClassifier(n_estimators=100, max_depth=20, random_state=42, n_jobs=-1)
clf_gs.fit(emb_tr_d0_s, tr_lbl_d0)

pred_orig    = clf_gs.predict(emb_orig_s)
pred_swapped = clf_gs.predict(emb_swapped_s)

# Flip rate global
flip_rate = (pred_orig != pred_swapped).mean()

# Flip rate par genre
print("\n" + "=" * 55)
print("GENDER SWAP — RÉSULTATS")
print("=" * 55)
print(f"  Flip rate global     : {flip_rate*100:.2f}%")
for g in sorted(val_gender_gs.unique()):
    mask_g = (val_gender_gs == g).values
    fr_g   = (pred_orig[mask_g] != pred_swapped[mask_g]).mean()
    print(f"  Flip rate ({g:6s})  : {fr_g*100:.2f}%")

# Perte de Macro-F1 après perturbation
mf1_orig    = f1_score(val_lbl_gs, pred_orig,    average='macro', zero_division=0)
mf1_swapped = f1_score(val_lbl_gs, pred_swapped, average='macro', zero_division=0)
print(f"\n  Macro-F1 original    : {mf1_orig*100:.2f}%")
print(f"  Macro-F1 après swap  : {mf1_swapped*100:.2f}%")
print(f"  Perte Macro-F1       : {(mf1_orig - mf1_swapped)*100:.2f} pts")
print(f"\n  → Un flip rate élevé indique que le modèle exploite les indices de genre.")


── Exemple gender swap ──
ORIGINAL : Danielle Daughtridge is a certified Physician Assistant who specializes in medical and cosmetic dermatology. She is a member of the American Academy of Physician Assistants, the New York Society of Ph
SWAPPED  : Danielle Daughtridge is a certified Physician Assistant who specializes in medical and cosmetic dermatology. He is a member of the American Academy of Physician Assistants, the New York Society of Phy

Extraction embeddings originaux (1000 textes)...
    480/1000
    960/1000
Extraction embeddings swappés  (1000 textes)...
    480/1000
    960/1000
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000

GENDER SWAP — RÉSULTATS
  Flip rate global     : 12.10%
  Flip rate (Female)  : 11.88%
  Flip rate (Male  )  : 12.39%

  Macro-F1 original    : 85.95%
  Macro-F1 après swap  : 8

---
## Interprétabilité — LIME

LIME (Local Interpretable Model-agnostic Explanations) perturbe localement le texte et entraîne un modèle linéaire pour expliquer la prédiction.  
On l'applique sur quelques biographies issues de D0 et D1n pour visualiser quels mots orientent la décision, avant et après nettoyage.


In [14]:
# === Installation LIME (si nécessaire) ===
subprocess.run([sys.executable, "-m", "pip", "install", "lime", "--quiet"], check=True)
print("LIME prêt.")


LIME prêt.


In [16]:
# === LIME — Interprétabilité sur D0 et D1n ===
from lime.lime_text import LimeTextExplainer

# ── Pipeline de prédiction BERT+RF ────────────────────────────────
# Utilise le modèle BiomedBERT + RF entraîné sur D0 (clf_gs / sc_gs)

def predict_proba_pipeline(texts, model=mdl_main, tokenizer=tok_main,
                            clf=clf_gs, scaler=sc_gs, dev=device):
    """Retourne les probabilités par classe pour LIME."""
    model.eval()
    all_emb = []
    batch_size = 8
    for i in range(0, len(texts), batch_size):
        batch = texts[i:i + batch_size]
        inputs = tokenizer(batch, padding=True, truncation=True,
                           max_length=512, return_tensors="pt").to(dev)
        with torch.no_grad():
            out = model(**inputs)
            all_emb.append(out.last_hidden_state[:, 0, :].cpu().numpy())
    emb = np.concatenate(all_emb, axis=0)
    emb_s = scaler.transform(emb)
    return clf.predict_proba(emb_s)

class_names = list(le_d.classes_)
explainer = LimeTextExplainer(class_names=class_names, random_state=42)

N_EXAMPLES = 3   # nombre de biographies à expliquer

print("=" * 60)
print("LIME — Mots importants sur D0 (texte brut)")
print("=" * 60)

# Sélectionner des exemples variés (un par profession si possible)
chosen_idxs = []
seen_labels = set()
for i, row in d0_val.iterrows():
    if row['label'] not in seen_labels:
        chosen_idxs.append(i)
        seen_labels.add(row['label'])
    if len(chosen_idxs) >= N_EXAMPLES:
        break

lime_results_d0  = []
lime_results_d1n = []

for idx in chosen_idxs:
    text_d0  = d0_val.loc[idx, 'full_text']
    text_d1n = d1n_val.loc[idx, 'full_text']
    true_lbl = d0_val.loc[idx, 'label']
    prof_name = le_d.classes_[true_lbl]

    # Prédiction
    pred_label_d0 = clf_gs.predict(
        sc_gs.transform(cls_embeddings(mdl_main, tok_main, [text_d0], device))
    )[0]

    print(f"\n── Exemple (vrai label : {prof_name}) ──")
    print(f"  Prédit : {le_d.classes_[pred_label_d0]}")
    print(f"  Texte  : {text_d0[:200]}...")

    # LIME sur D0
    exp_d0 = explainer.explain_instance(
        text_d0,
        predict_proba_pipeline,
        num_features=10,
        labels=[true_lbl],
        num_samples=300,
    )
    print(f"\n  Top mots importants (D0) pour la classe '{prof_name}':")
    for word, weight in exp_d0.as_list(label=true_lbl)[:10]:
        sign = "+" if weight > 0 else "-"
        print(f"    {sign} {word:30s}  {weight:.4f}")

    # LIME sur D1n (même exemple, texte nettoyé)
    exp_d1n = explainer.explain_instance(
        text_d1n,
        predict_proba_pipeline,
        num_features=10,
        labels=[true_lbl],
        num_samples=300,
    )
    print(f"\n  Top mots importants (D1n) pour la classe '{prof_name}':")
    for word, weight in exp_d1n.as_list(label=true_lbl)[:10]:
        sign = "+" if weight > 0 else "-"
        print(f"    {sign} {word:30s}  {weight:.4f}")

    lime_results_d0.append(exp_d0)
    lime_results_d1n.append(exp_d1n)

print("\nAnalyse LIME terminée.")


LIME — Mots importants sur D0 (texte brut)

── Exemple (vrai label : psychologist) ──
  Prédit : psychologist
  Texte  : Adam Grant is an organizational psychologist who has been repeatedly recognized as Wharton University's top-rated professor (although he's only 35 years old). He's also written multiple New York Times...

  Top mots importants (D0) pour la classe 'psychologist':
    + psychologist                    0.2155
    + s                               0.0784
    + He                              0.0752
    + organizational                  0.0723
    + scientific                      0.0424
    + giving                          0.0402
    + Give                            0.0390
    + leads                           0.0382
    - 35                              -0.0201
    - rated                           -0.0128

  Top mots importants (D1n) pour la classe 'psychologist':
    + workplace                       0.0360
    + giving                          0.0349
    + for     

---
## Tableau final comparatif — BiomedBERT frozen + RF

| Modèle | Données | Accuracy | Macro-F1 | DP gap | EOpp gap | FPR gap | Flip rate |
|--------|---------|----------|----------|--------|----------|---------|-----------|
| BiomedBERT frozen + RF | D0 | 87.20 % | 85.95 % | 0.00 % | 8.80 % | 1.05 % | 12.10 % |
| BiomedBERT frozen + RF | D1 | 78.40 % | 76.97 % | 0.00 % | 4.45 % | 1.30 % | — |
| BiomedBERT frozen + RF | D1n | 77.50 % | 76.26 % | 0.00 % | 6.11 % | 1.36 % | — |

*(Généré par la cellule suivante — résultats complets 6 modèles dans la cellule 38)*


In [18]:
# === Tableau final comparatif ===
MODEL_SHORT = "BiomedBERT frozen + RF"

rows = []
for version in ["D0", "D1", "D1n"]:
    if version not in fairness_results:
        continue
    r      = fairness_results[version]
    fd     = r['fairness']
    y_true = r['y_true']
    y_pred = r['y_pred']

    acc  = accuracy_score(y_true, y_pred)
    mf1  = f1_score(y_true, y_pred, average='macro',    zero_division=0)
    wf1  = f1_score(y_true, y_pred, average='weighted', zero_division=0)

    row = {
        "Modèle":    MODEL_SHORT,
        "Données":   version,
        "Accuracy":  f"{acc*100:.2f}%",
        "Macro-F1":  f"{mf1*100:.2f}%",
        "Wgt-F1":    f"{wf1*100:.2f}%",
        "DP gap":    f"{fd['dp_gap']*100:.2f}%",
        "EOpp gap":  f"{fd['eopp_gap']*100:.2f}%",
        "FPR gap":   f"{fd['fpr_gap']*100:.2f}%",
        "Flip rate": "—",
    }
    rows.append(row)

# Ajouter la ligne gender swap (D0 uniquement, calculé plus haut)
if 'flip_rate' in dir():
    for row in rows:
        if row["Données"] == "D0":
            row["Flip rate"] = f"{flip_rate*100:.2f}%"

df_final = pd.DataFrame(rows)
print("=" * 90)
print("TABLEAU FINAL COMPARATIF")
print("=" * 90)
print(df_final.to_string(index=False))

print("\n── Analyse des compromis ──")
print("  D0  : performance maximale mais fuite de label (profession dans le texte).")
print("  D1  : suppression des mots-métiers → évaluation plus réaliste, fairness gap réduit.")
print("  D1n : D1 + neutralisation genre → biais de genre réduit, légère perte d'accuracy possible.")
print("  Flip rate : mesure la sensibilité du modèle aux indices de genre dans D0.")


TABLEAU FINAL COMPARATIF
                Modèle Données Accuracy Macro-F1 Wgt-F1 DP gap EOpp gap FPR gap Flip rate
BiomedBERT frozen + RF      D0   87.20%   85.95% 86.88%  0.00%    8.80%   1.05%    12.10%
BiomedBERT frozen + RF      D1   78.40%   76.97% 77.94%  0.00%    4.45%   1.30%         —
BiomedBERT frozen + RF     D1n   77.50%   76.26% 77.17%  0.00%    6.11%   1.36%         —

── Analyse des compromis ──
  D0  : performance maximale mais fuite de label (profession dans le texte).
  D1  : suppression des mots-métiers → évaluation plus réaliste, fairness gap réduit.
  D1n : D1 + neutralisation genre → biais de genre réduit, légère perte d'accuracy possible.
  Flip rate : mesure la sensibilité du modèle aux indices de genre dans D0.


---
## Comparaison complète : 6 modèles (5 biomédicaux + DistilBERT baseline) × D0 / D1 / D1n

| Modèle | D0 Macro-F1 | D1 Macro-F1 | D1n Macro-F1 | D1n EOpp gap |
|--------|------------|------------|-------------|-------------|
| biomed_roberta_base | **87.37 %** | **82.86 %** | **81.71 %** | 8.34 % |
| BiomedBERT-base-uncased | 85.95 % | 76.97 % | 76.26 % | 6.11 % |
| distilbert-base-uncased | 87.36 % | 80.57 % | 78.47 % | 13.60 % ❌ |
| Bio_ClinicalBERT | 77.50 % | 72.16 % | 70.70 % | **3.92 %** ✅ |
| biobert-base-cased-v1.2 | 77.49 % | 70.29 % | 65.90 % | 8.49 % |
| BioM-ELECTRA-Large | 63.70 % | 66.57 % | 60.12 % | 11.64 % |

*(Tableau détaillé généré par la cellule suivante)*


In [6]:
# === Pipeline fairness complet : 6 modèles × 3 versions de données ===
import gc

ALL_MODELS = [
    "distilbert-base-uncased",
    "dmis-lab/biobert-base-cased-v1.2",
    "emilyalsentzer/Bio_ClinicalBERT",
    "allenai/biomed_roberta_base",
    "sultan/BioM-ELECTRA-Large-SQuAD2",
    "microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract",
]

DATASETS_MULTI = {
    "D0":  (d0_train, d0_val),
    "D1":  (d1_train, d1_val),
    "D1n": (d1n_train, d1n_val),
}

classes_multi = list(range(len(le_d.classes_)))
all_results = []

for model_id in ALL_MODELS:
    short = model_id.split("/")[-1]
    print(f"\n{'='*65}")
    print(f"MODÈLE : {model_id}")
    print(f"{'='*65}")

    bs = 8 if "large" in model_id.lower() else 16

    try:
        tok = AutoTokenizer.from_pretrained(model_id)
        mdl = AutoModel.from_pretrained(model_id).to(device)
        mdl.eval()

        for version, (tr_df, va_df) in DATASETS_MULTI.items():
            print(f"\n  ── Version {version} ──")
            tr_texts  = tr_df['full_text'].tolist()
            va_texts  = va_df['full_text'].tolist()
            tr_lbl    = tr_df['label'].values
            va_lbl    = va_df['label'].values
            va_gender = va_df['gender']

            tr_emb = cls_embeddings(mdl, tok, tr_texts, device, batch_size=bs)
            va_emb = cls_embeddings(mdl, tok, va_texts, device, batch_size=bs)

            sc = StandardScaler()
            tr_s = sc.fit_transform(tr_emb)
            va_s = sc.transform(va_emb)

            clf = RandomForestClassifier(n_estimators=100, max_depth=20, random_state=42, n_jobs=-1)
            clf.fit(tr_s, tr_lbl)
            va_pred = clf.predict(va_s)

            acc = accuracy_score(va_lbl, va_pred)
            mf1 = f1_score(va_lbl, va_pred, average='macro',    zero_division=0)
            wf1 = f1_score(va_lbl, va_pred, average='weighted', zero_division=0)
            fd  = compute_fairness_metrics(va_lbl, va_pred, va_gender, classes_multi)

            print(f"    Acc={acc*100:.2f}%  Macro-F1={mf1*100:.2f}%  "
                  f"EOpp gap={fd['eopp_gap']*100:.2f}%  DP gap={fd['dp_gap']*100:.2f}%")

            all_results.append({
                "Modèle":   short,
                "Données":  version,
                "Accuracy": acc,
                "Macro-F1": mf1,
                "Wgt-F1":   wf1,
                "DP gap":   fd['dp_gap'],
                "EOpp gap": fd['eopp_gap'],
                "FPR gap":  fd['fpr_gap'],
                "Acc gap":  fd['acc_gap'],
            })

            del clf, sc, tr_emb, va_emb, tr_s, va_s

    except Exception as e:
        print(f"  ERREUR : {e}")
        for version in DATASETS_MULTI:
            all_results.append({
                "Modèle": short, "Données": version,
                "Accuracy": None, "Macro-F1": None, "Wgt-F1": None,
                "DP gap": None, "EOpp gap": None, "FPR gap": None, "Acc gap": None,
            })

    finally:
        try: del mdl, tok
        except: pass
        gc.collect()
        if torch.cuda.is_available():
            torch.cuda.empty_cache()

# ── Tableau récapitulatif ──────────────────────────────────────────
df_multi = pd.DataFrame(all_results)

def fmt(v):
    if v is None: return "ERR"
    return f"{v*100:.2f}%"

print("\n\n" + "=" * 115)
print("TABLEAU COMPARATIF COMPLET — 6 modèles × D0/D1/D1n")
print("=" * 115)

df_display = df_multi.copy()
for col in ["Accuracy", "Macro-F1", "Wgt-F1", "DP gap", "EOpp gap", "FPR gap", "Acc gap"]:
    df_display[col] = df_display[col].apply(fmt)

print(df_display.to_string(index=False))

# ── Meilleur modèle par version (Macro-F1) ────────────────────────
print("\n── Meilleur modèle par version (Macro-F1) ──")
for version in ["D0", "D1", "D1n"]:
    subset = df_multi[df_multi["Données"] == version].dropna(subset=["Macro-F1"])
    if not subset.empty:
        best = subset.loc[subset["Macro-F1"].idxmax()]
        print(f"  {version}: {best['Modèle']:50s}  Macro-F1={best['Macro-F1']*100:.2f}%  EOpp gap={best['EOpp gap']*100:.2f}%")

# ── Modèle le plus équitable sur D1n (EOpp gap minimal) ───────────
print("\n── Modèle le plus équitable sur D1n (EOpp gap minimal) ──")
d1n_sub = df_multi[df_multi["Données"] == "D1n"].dropna(subset=["EOpp gap"])
if not d1n_sub.empty:
    fairest = d1n_sub.loc[d1n_sub["EOpp gap"].idxmin()]
    print(f"  {fairest['Modèle']}  EOpp gap={fairest['EOpp gap']*100:.2f}%  Macro-F1={fairest['Macro-F1']*100:.2f}%")



MODÈLE : distilbert-base-uncased


config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertModel LOAD REPORT from: distilbert-base-uncased
Key                     | Status     |  | 
------------------------+------------+--+-
vocab_transform.weight  | UNEXPECTED |  | 
vocab_projector.bias    | UNEXPECTED |  | 
vocab_layer_norm.weight | UNEXPECTED |  | 
vocab_layer_norm.bias   | UNEXPECTED |  | 
vocab_transform.bias    | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.



  ── Version D0 ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=88.50%  Macro-F1=87.36%  EOpp gap=4.11%  DP gap=0.00%

  ── Version D1 ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=81.60%  Macro-F1=80.57%  EOpp gap=7.25%  DP gap=0.00%

  ── Version D1n ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=79.70%  Macro-F1=78.47%  EOpp gap=13.60%  DP gap=0.00%

MODÈLE 

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: dmis-lab/biobert-base-cased-v1.2
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.predictions.decoder.bias               | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.predictions.decoder.weight             | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.



  ── Version D0 ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=78.20%  Macro-F1=77.49%  EOpp gap=8.47%  DP gap=0.00%

  ── Version D1 ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=71.30%  Macro-F1=70.29%  EOpp gap=9.76%  DP gap=0.00%

  ── Version D1n ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=67.20%  Macro-F1=65.90%  EOpp gap=8.49%  DP gap=0.00%

MODÈLE :

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: emilyalsentzer/Bio_ClinicalBERT
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.predictions.decoder.weight             | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.



  ── Version D0 ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=78.50%  Macro-F1=77.50%  EOpp gap=4.20%  DP gap=0.00%

  ── Version D1 ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=72.70%  Macro-F1=72.16%  EOpp gap=5.66%  DP gap=0.00%

  ── Version D1n ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=72.00%  Macro-F1=70.70%  EOpp gap=3.92%  DP gap=0.00%

MODÈLE :

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

RobertaModel LOAD REPORT from: allenai/biomed_roberta_base
Key                       | Status     |  | 
--------------------------+------------+--+-
lm_head.dense.bias        | UNEXPECTED |  | 
lm_head.layer_norm.bias   | UNEXPECTED |  | 
lm_head.dense.weight      | UNEXPECTED |  | 
lm_head.decoder.weight    | UNEXPECTED |  | 
lm_head.layer_norm.weight | UNEXPECTED |  | 
lm_head.bias              | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.



  ── Version D0 ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=88.50%  Macro-F1=87.37%  EOpp gap=6.78%  DP gap=0.00%

  ── Version D1 ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=83.60%  Macro-F1=82.86%  EOpp gap=3.79%  DP gap=0.00%

  ── Version D1n ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=82.90%  Macro-F1=81.71%  EOpp gap=8.34%  DP gap=0.00%

MODÈLE :

Loading weights:   0%|          | 0/389 [00:00<?, ?it/s]

ElectraModel LOAD REPORT from: sultan/BioM-ELECTRA-Large-SQuAD2
Key                             | Status     |  | 
--------------------------------+------------+--+-
qa_outputs.weight               | UNEXPECTED |  | 
electra.embeddings.position_ids | UNEXPECTED |  | 
qa_outputs.bias                 | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.



  ── Version D0 ──
    240/8000
    480/8000
    720/8000
    960/8000
    1200/8000
    1440/8000
    1680/8000
    1920/8000
    2160/8000
    2400/8000
    2640/8000
    2880/8000
    3120/8000
    3360/8000
    3600/8000
    3840/8000
    4080/8000
    4320/8000
    4560/8000
    4800/8000
    5040/8000
    5280/8000
    5520/8000
    5760/8000
    6000/8000
    6240/8000
    6480/8000
    6720/8000
    6960/8000
    7200/8000
    7440/8000
    7680/8000
    7920/8000
    240/1000
    480/1000
    720/1000
    960/1000
    Acc=64.90%  Macro-F1=63.70%  EOpp gap=8.87%  DP gap=0.00%

  ── Version D1 ──
    240/8000
    480/8000
    720/8000
    960/8000
    1200/8000
    1440/8000
    1680/8000
    1920/8000
    2160/8000
    2400/8000
    2640/8000
    2880/8000
    3120/8000
    3360/8000
    3600/8000
    3840/8000
    4080/8000
    4320/8000
    4560/8000
    4800/8000
    5040/8000
    5280/8000
    5520/8000
    5760/8000
    6000/8000
    6240/8000
    6480/8000
    6720/8000


Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.predictions.decoder.bias               | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.predictions.decoder.weight             | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.



  ── Version D0 ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=87.20%  Macro-F1=85.95%  EOpp gap=8.80%  DP gap=0.00%

  ── Version D1 ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=78.40%  Macro-F1=76.97%  EOpp gap=4.45%  DP gap=0.00%

  ── Version D1n ──
    480/8000
    960/8000
    1440/8000
    1920/8000
    2400/8000
    2880/8000
    3360/8000
    3840/8000
    4320/8000
    4800/8000
    5280/8000
    5760/8000
    6240/8000
    6720/8000
    7200/8000
    7680/8000
    480/1000
    960/1000
    Acc=77.50%  Macro-F1=76.26%  EOpp gap=6.11%  DP gap=0.00%


TABLEAU